scipy/stats/tests/test_qmc.py

```diff
@@ -0,0 +1,868 @@
+                           [2, 2, 2, 1, 2, 2]])
+        sample = (2.0 * sample - 1.0) / (2.0 * 2.0)
+
+        assert_allclose(qmc.discrepancy(sample, method='MD'), 2.5000,
+                        atol=1e-4)
+        assert_allclose(qmc.discrepancy(sample, method='WD'), 1.3680,
+                        atol=1e-4)
+        assert_allclose(qmc.discrepancy(sample, method='CD'), 0.3172,
+                        atol=1e-4)
+        assert_allclose(qmc.discrepancy(sample, method='star'), 0.037451,
+                        atol=1e-4)
+
+        with pytest.raises(ValueError, match=r"toto is not a valid method."):
+            qmc.discrepancy(sample, method='toto')
+
+    def test_update_discrepancy(self):
+        space_1 = np.array([[1, 3], [2, 6], [3, 2], [4, 5], [5, 1], [6, 4]])
+        space_1 = (2.0 * space_1 - 1.0) / (2.0 * 6.0)
+
+        disc_init = qmc.discrepancy(space_1[:-1], iterative=True)
+        disc_iter = _update_discrepancy(space_1[-1], space_1[:-1],
+                                        disc_init)
+
+        assert_allclose(disc_iter, 0.0081, atol=1e-4)
+
+    def test_perm_discrepancy(self):
+        doe_init = np.array([[1, 3], [2, 6], [3, 2], [4, 5], [5, 1], [6, 4]])
+        doe_init = (2.0 * doe_init - 1.0) / (2.0 * 6.0)
+
+        disc_init = qmc.discrepancy(doe_init)
+
+        row_1, row_2, col = 5, 2, 1
+
+        doe = copy.deepcopy(doe_init)
+        doe[row_1, col], doe[row_2, col] = doe[row_2, col], doe[row_1, col]
+
+        disc_valid = qmc.discrepancy(doe)
+        disc_perm = _perturb_discrepancy(doe_init, row_1, row_2, col,
+                                         disc_init)
+
+        assert_allclose(disc_valid, disc_perm)
+
+    def test_n_primes(self):
+        primes = n_primes(10)
+        assert primes[-1] == 29
+
+        primes = n_primes(168)
+        assert primes[-1] == 997
+
+        primes = n_primes(350)
+        assert primes[-1] == 2357
+
+    def test_primes(self):
+        primes = primes_from_2_to(50)
+        out = [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47]
+        assert_allclose(primes, out)
+
+
+class TestVDC:
+    def test_van_der_corput(self):
+        seed = np.random.RandomState(12345)
+        sample = van_der_corput(10, seed=seed)
+        out = [0., 0.5, 0.25, 0.75, 0.125, 0.625, 0.375, 0.875, 0.0625, 0.5625]
+        assert_almost_equal(sample, out)
+
+        sample = van_der_corput(7, start_index=3, seed=seed)
+        assert_almost_equal(sample, out[3:])
+
+    def test_van_der_corput_scramble(self):
+        seed = np.random.RandomState(123456)
+        out = van_der_corput(10, scramble=True, seed=seed)
+
+        seed = np.random.RandomState(123456)
+        sample = van_der_corput(7, start_index=3, scramble=True,
+                                seed=seed)
+        assert_almost_equal(sample, out[3:])
+
+
+class QMCEngineTests:
+    """Generic tests for QMC engines."""
+    qmce = ...
+    can_scramble = ...
+    unscramble_nd = ...
+    scramble_nd = ...
```

Comment:
These ellipses are a little odd, and mypy doesn't like them. Adding `# type: ignore` doesn't help. See ABC review comment.

Reply:
In fact, these ellipsis are causing most of the mypy errors observed in https://github.com/scipy/scipy/pull/10844#pullrequestreview-560646097; because of them the type of these attributes is inferred as `ellipsis`.

I'd suggest either replacing them with `NotImplemented` (which is treated as `Any` by static type checkers) or removing them altogether.

Reply:
Ok thanks, I will change this to `NotImplemented` as you suggested. I prefer this over removing them.